File: java/3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) return 0;
        int max = 1;
        int i = 0, j = -1;
        int[] lookup = new int[256];
        Arrays.fill(lookup, -1);

        while (j < s.length() - 1) {
            if (lookup[s.charAt(++j)] >= i)
                i = lookup[s.charAt(j)] + 1;
            
            lookup[s.charAt(j)] = j;
            if (max < j-i+1) max = j-i+1;
        }
        return max;
    }
}

File: go/3.longest-substring-without-repeating-characters.go
/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 {
        return 0
    }
    max, start := 1, 0
    m := make(map[byte]int)
    for i := 0; i < len(s); i++ {
        c := s[i]
        if pos, b := m[c]; b && start <= pos {
            start = pos + 1
        }
        m[c] = i
        
        if max < i - start + 1 {
            max = i - start + 1
        }
    }
    return max
}

// @lc code=end

